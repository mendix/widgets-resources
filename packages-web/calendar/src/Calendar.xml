<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.custom.calendar.Calendar" needsEntityContext="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/">
    <name>Calendar</name>
    <description>Display and manage calendar events</description>
    <icon/>
    <properties>
        <property key="defaultView" type="enumeration" defaultValue="month">
            <caption>Default view</caption>
            <category>Appearance</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="month">Month</enumerationValue>
                <enumerationValue key="week">Week</enumerationValue>
                <enumerationValue key="work_week">Work week</enumerationValue>
                <enumerationValue key="day">Day</enumerationValue>
                <enumerationValue key="agenda">Agenda</enumerationValue>
            </enumerationValues>
        </property>
        <property key="popup" type="boolean" defaultValue="true">
            <caption>Popup</caption>
            <category>Appearance</category>
            <description>Show truncated events in an overlay when you click the "+x more" link</description>
        </property>
        <property key="selectable" type="boolean" defaultValue="true">
            <caption>Selectable</caption>
            <category>Appearance</category>
            <description/>
        </property>
        <property key="editable" type="enumeration" defaultValue="default">
            <caption>Editable</caption>
            <category>Appearance</category>
            <description>The default option is based on the page settings and the attribute's write access</description>
            <enumerationValues>
                <enumerationValue key="default">Default</enumerationValue>
                <enumerationValue key="never">Never</enumerationValue>
            </enumerationValues>
        </property>
        <property key="dataSource" type="enumeration" defaultValue="XPath">
            <caption>Data source</caption>
            <category>Event data</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="XPath">Database</enumerationValue>
                <enumerationValue key="microflow">Microflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="eventEntity" type="entity" required="false">
            <caption>Event entity</caption>
            <category>Event data</category>
            <description/>
        </property>
        <property key="entityConstraint" type="entityConstraint" entityProperty="eventEntity" required="false">
            <caption>XPath</caption>
            <category>Event data</category>
            <description/>
        </property>
		<property key="dataSourceMicroflow" type="microflow" required="false">
			<caption>Data source microflow</caption>
			<category>Event data</category>
			<description></description>
			<returnType type="Void" />
		</property>
        <property key="titleAttribute" type="attribute" entityProperty="eventEntity" isPath="optional" pathType="reference" required="false">
            <caption>Title</caption>
            <category>Event data</category>
            <description/>
            <attributeTypes>
                <attributeType name="String"/>
            </attributeTypes>
        </property>
        <property key="startAttribute" type="attribute" entityProperty="eventEntity" required="false">
            <caption>Start</caption>
            <category>Event data</category>
            <description/>
            <attributeTypes>
                <attributeType name="DateTime"/>
            </attributeTypes>
        </property>
        <property key="endAttribute" type="attribute" entityProperty="eventEntity" required="false">
            <caption>End</caption>
            <category>Event data</category>
            <description/>
            <attributeTypes>
                <attributeType name="DateTime"/>
            </attributeTypes>
        </property>
        <property key="onClickEvent" type="enumeration" defaultValue="doNothing">
            <caption>On click event</caption>
            <category>Events</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="doNothing">Do nothing</enumerationValue>
                <enumerationValue key="showPage">Show a page</enumerationValue>
                <enumerationValue key="callMicroflow">Call a microflow</enumerationValue>
                <enumerationValue key="callNanoflow">Call a nanoflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="eventPage" type="form" required="false">
            <caption>Page</caption>
            <category>Events</category>
            <description>The page that will be opened on clicking an event</description>
        </property>
        <property key="openPageAs" type="enumeration" defaultValue="content">
            <caption>Open page as</caption>
            <category>Events</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="content">Full page</enumerationValue>
                <enumerationValue key="popup">Popup</enumerationValue>
                <enumerationValue key="modal">Blocking popup</enumerationValue>
            </enumerationValues>
        </property>
        <property key="eventMicroflow" type="microflow" required="false">
            <caption>Microflow</caption>
            <category>Events</category>
            <description>The microflow that will be executed on clicking an event</description>
            <returnType type="Void" />
        </property>
        <property key="eventNanoflow" type="nanoflow" required="false">
            <caption>Nanoflow</caption>
            <category>Events</category>
            <description>The nanoflow that will be executed on clicking an event</description>
            <returnType type="Void" />
        </property>
        <property key="onClickSlotEvent" type="enumeration" defaultValue="doNothing">
            <caption>On click slot</caption>
            <category>Events</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="doNothing">Do nothing</enumerationValue>
                <enumerationValue key="showPage">Show a page</enumerationValue>
                <enumerationValue key="callMicroflow">Call a microflow</enumerationValue>
                <enumerationValue key="callNanoflow">Call a nanoflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="slotPage" type="form" required="false">
            <caption>New event page</caption>
            <category>Events</category>
            <description>The page that will be opened on clicking a slot</description>
        </property>
        <property key="slotMicroflow" type="microflow" required="false">
            <caption>New event microflow</caption>
            <category>Events</category>
            <description>The microflow that will be executed on clicking a slot</description>
            <returnType type="Void" />
        </property>
        <property key="slotNanoflow" type="nanoflow" required="false">
            <caption>New event nanoflow</caption>
            <category>Events</category>
            <description>The nanoflow that will be executed on clicking a slot</description>
            <returnType type="Void" />
        </property>
        <property key="onDropEvent" type="enumeration" defaultValue="doNothing">
            <caption>On drop event</caption>
            <category>Events</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="doNothing">Do nothing</enumerationValue>
                <enumerationValue key="showPage">Show a page</enumerationValue>
                <enumerationValue key="callMicroflow">Call a microflow</enumerationValue>
                <enumerationValue key="callNanoflow">Call a nanoflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="onDropPage" type="form" required="false">
            <caption>On drop page</caption>
            <category>Events</category>
            <description>The page that will be opened on droping an event</description>
        </property>
        <property key="onDropMicroflow" type="microflow" required="false">
            <caption>On drop microflow</caption>
            <category>Events</category>
            <description>The microflow that will be executed on droping an event</description>
            <returnType type="Void" />
        </property>
        <property key="onDropNanoflow" type="nanoflow" required="false">
            <caption>On drop nanoflow</caption>
            <category>Events</category>
            <description>The nanoflow that will be executed on droping an event</description>
            <returnType type="Void" />
        </property>
    </properties>
</widget>
