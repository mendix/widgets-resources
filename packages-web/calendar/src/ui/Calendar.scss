.widget-calendar {
    .rbc-calendar {
        min-height: 580px !important;
        height: 580px;
    }
}
