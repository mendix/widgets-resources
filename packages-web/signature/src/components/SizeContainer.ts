import { CSSProperties, SFC, createElement } from "react";
import classNames from "classnames";

export type HeightUnitType = "percentageOfWidth" | "percentageOfParent" | "pixels";

export type WidthUnitType = "percentage" | "pixels";

export interface Dimensions {
    widthUnit: WidthUnitType;
    width: number;
    heightUnit: HeightUnitType;
    height: number;
}

export interface SizeProps extends Dimensions {
    className: string;
    classNameInner?: string;
    readOnly?: boolean;
    style?: CSSProperties;
}

export const SizeContainer: SFC<SizeProps> = ({
    className,
    classNameInner,
    widthUnit,
    width,
    heightUnit,
    height,
    children,
    style,
    readOnly
}) => {
    const styleWidth = widthUnit === "percentage" ? `${width}%` : `${width}px`;
    return createElement(
        "div",
        {
            className: classNames(className, "size-box"),
            style: {
                position: "relative",
                width: styleWidth,
                ...getHeight(heightUnit, height, widthUnit, width),
                ...style
            }
        },
        createElement(
            "div",
            {
                className: classNames("size-box-inner", classNameInner),
                readOnly,
                disabled: readOnly,
                style: {
                    position: "absolute",
                    top: "0",
                    right: "0",
                    bottom: "0",
                    left: "0"
                }
            },
            children
        )
    );
};

SizeContainer.displayName = "SizeContainer";
SizeContainer.defaultProps = { readOnly: false };

const getHeight = (
    heightUnit: HeightUnitType,
    height: number,
    widthUnit: WidthUnitType,
    width: number
): CSSProperties => {
    const style: CSSProperties = {};
    if (heightUnit === "percentageOfWidth") {
        const ratio = (height / 100) * width;
        if (widthUnit === "percentage") {
            style.height = "auto";
            style.paddingBottom = `${ratio}%`;
        } else {
            style.height = `${ratio}px`;
        }
    } else if (heightUnit === "pixels") {
        style.height = `${height}px`;
    } else if (heightUnit === "percentageOfParent") {
        style.height = `${height}%`;
    }

    return style;
};
