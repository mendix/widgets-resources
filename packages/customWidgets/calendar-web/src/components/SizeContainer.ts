import { CSSProperties, createElement, FunctionComponent } from "react";
import classNames from "classnames";

export type HeightUnitType = "percentageOfWidth" | "percentageOfParent" | "pixels";

export type WidthUnitType = "percentage" | "pixels";

export interface Dimensions {
    widthUnit: WidthUnitType;
    width: number;
    heightUnit: HeightUnitType;
    height: number;
}

export interface SizeProps extends Dimensions {
    className: string;
    style?: CSSProperties;
}

export const SizeContainer: FunctionComponent<SizeProps> = ({
    className,
    widthUnit,
    width,
    heightUnit,
    height,
    children,
    style
}) => {
    const styleWidth = widthUnit === "percentage" ? `${width}%` : `${width}px`;

    return createElement(
        "div",
        {
            className: classNames(className, "size-box"),
            style: {
                position: "relative",
                width: styleWidth,
                ...getHeight(heightUnit, height),
                ...style
            },
            ref: parentHeight
        },
        createElement(
            "div",
            {
                className: "size-box-inner",
                style: {
                    position: "absolute",
                    top: "0",
                    right: "0",
                    bottom: "0",
                    left: "0"
                }
            },
            children
        )
    );
};

const parentHeight = (node?: HTMLElement | null): void => {
    // Fix for percentage height of parent.
    // There no other way to control widget wrapper style
    if (node && node.parentElement) {
        node.parentElement.style.height = "100%";
    }
};

const getHeight = (heightUnit: HeightUnitType, height: number): CSSProperties => {
    const style: CSSProperties = {};

    if (heightUnit === "percentageOfWidth") {
        style.height = "auto";
        style.paddingBottom = `${height}%`;
    } else if (heightUnit === "pixels") {
        style.height = `${height}px`;
    } else if (heightUnit === "percentageOfParent") {
        style.height = `${height}%`;
    }

    return style;
};
