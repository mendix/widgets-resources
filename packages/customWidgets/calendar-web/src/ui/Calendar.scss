@import "variables";

.widget-calendar {
    min-height: 350px;
    margin-bottom: $form-group-margin-bottom;
    user-select: none;
    -ms-user-select: none;

    .rbc-calendar {
        width: 100%;
        height: 100%;
    }

    .rbc-event {
        min-height: 20px;
    }

    .rbc-addons-dnd .rbc-addons-dnd-resize-ns-icon {
        cursor: row-resize;
    }

    .rbc-addons-dnd .rbc-addons-dnd-resize-ew-icon {
        cursor: col-resize;
    }

    .calendar-label {
        font-size: 16px;
    }

    .calendar-toolbar {
        display: flex;
        align-items: center;
        justify-content: space-between;
        flex-wrap: wrap;
        padding-bottom: 10px;
    }

    .rbc-current-time-indicator {
        background-color: $brand-danger;
    }

    .rbc-day-slot .rbc-event {
        border-top-color: $border-color-default;
        border-right-color: $border-color-default;
        border-bottom-color: $border-color-default;
        border-left-color: $border-color-default;
    }

    .rbc-off-range-bg {
        background-color: $color-default-lighter;
    }

    .rbc-today {
        background-color: $color-primary-lighter;
    }

    .rbc-event {
        background-color: $brand-primary;
    }
}
