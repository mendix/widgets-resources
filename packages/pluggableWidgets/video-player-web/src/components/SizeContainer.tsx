import { CSSProperties, Component, createElement } from "react";
import classNames from "classnames";

export type HeightUnitType = "percentageOfWidth" | "percentageOfParent" | "pixels" | "aspectRatio";

export type WidthUnitType = "percentage" | "pixels";

export interface Dimensions {
    widthUnit: WidthUnitType;
    width: number;
    heightUnit: HeightUnitType;
    height: number;
    tabIndex?: number;
}

export interface SizeProps extends Dimensions {
    className: string;
    classNameInner?: string;
    style?: CSSProperties;
}

export class SizeContainer extends Component<SizeProps> {
    render(): JSX.Element {
        const styleWidth = this.props.widthUnit === "percentage" ? `${this.props.width}%` : `${this.props.width}px`;
        const relativeStyle: CSSProperties = {
            position: "relative",
            width: styleWidth,
            ...this.getHeight(this.props.heightUnit, this.props.height, this.props.widthUnit, this.props.width),
            display: "flex",
            justifyContent: "center",
            ...this.props.style
        };
        const absoluteStyle: CSSProperties = {
            position: "absolute",
            top: "0",
            right: "0",
            bottom: "0",
            left: "0",
            display: "flex",
            justifyContent: "center"
        };

        return (
            <div
                className={classNames(this.props.className, "size-box")}
                style={relativeStyle}
                tabIndex={this.props.tabIndex}
            >
                <div className={classNames("size-box-inner", this.props.classNameInner)} style={absoluteStyle}>
                    {this.props.children}
                </div>
            </div>
        );
    }

    private getHeight(
        heightUnit: HeightUnitType,
        height: number,
        widthUnit: WidthUnitType,
        width: number
    ): CSSProperties {
        const style: CSSProperties = {};
        if (heightUnit === "percentageOfWidth") {
            const ratio = (height / 100) * width;
            if (widthUnit === "percentage") {
                style.height = "auto";
                style.paddingBottom = `${ratio}%`;
            } else {
                style.height = `${ratio}px`;
            }
        } else if (heightUnit === "pixels") {
            style.height = `${height}px`;
        } else if (heightUnit === "percentageOfParent") {
            style.height = `${height}%`;
        } else if (heightUnit === "aspectRatio") {
            style.height = width * 0.5625; // Default is 16:9
        }

        return style;
    }
}
