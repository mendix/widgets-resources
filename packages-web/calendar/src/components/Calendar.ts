import { CSSProperties, Component, ReactChild, createElement, ReactNode } from "react";

import { Alert } from "./Alert";
import { Container, Style } from "../utils/namespaces";
import classNames from "classnames";
import BigCalendar from "react-big-calendar";
import moment from "moment";
import withDragAndDrop from "react-big-calendar/lib/addons/dragAndDrop";
import CustomToolbar from "./Toolbar";
import { SizeContainer } from "./SizeContainer";

import { CalendarLoader } from "./CalendarLoader";

import "react-big-calendar/lib/addons/dragAndDrop/styles.css";
import "react-big-calendar/lib/css/react-big-calendar.css";
import "../ui/Calendar.scss";
import "../ui/CalendarLoader.scss";

const localizer = BigCalendar.momentLocalizer(moment);
const DragAndDropCalendar = withDragAndDrop(BigCalendar);

export interface CalendarProps {
    alertMessage?: ReactChild;
    className?: string;
    events: CalendarEvent[];
    color?: string;
    formats?: {};
    enableCreate: boolean;
    height: number;
    heightUnit: Style.HeightUnitType;
    defaultView: Style.View;
    loading?: boolean;
    startPosition?: Date;
    messages: {};
    popup: boolean;
    editable: string;
    titleFormat?: (date: Date) => void;
    weekdayFormat?: (date: Date) => void;
    timeGutterFormat?: (date: Date) => void;
    monthHeaderFormat?: (date: Date) => void;
    dayHeaderFormat?: (date: Date) => void;
    style: CSSProperties;
    viewOption: "custom" | "standard";
    width: number;
    widthUnit: Style.WidthUnitType;
    onRangeChangeAction?: (date: object) => void;
    onSelectEventAction?: (eventInfo: object) => void;
    onEventResizeAction?: (eventInfo: object) => void;
    onSelectSlotAction?: (slotInfo: object) => void;
    onEventDropAction?: (eventInfo: object) => void;
    customViews: Container.CustomViews[];
}

interface HOCToolbarProps {
    customViews: Container.CustomViews[] | Array<Partial<Container.ButtonConfig>>;
    onClickToolbarButton: (date: object) => void;
}

export interface CalendarEvent {
    allDay: boolean;
    color: string;
    end: Date;
    guid: string;
    start: Date;
    title: string;
}

interface State {
    date?: Date;
}

class Calendar extends Component<CalendarProps, State> {
    readonly state: State = {
        date: undefined
    };

    private readonly onNavigateHandler = this.onNavigate.bind(this);

    render(): ReactNode {
        return createElement(
            SizeContainer,
            {
                className: classNames("widget-calendar", this.props.className),
                style: this.props.style,
                widthUnit: this.props.widthUnit,
                width: this.props.width,
                heightUnit: this.props.heightUnit,
                height: this.props.height
            },
            this.renderAlert(),
            this.renderCalendar()
        );
    }

    private getDefaultToolbar(): Array<Partial<Container.ButtonConfig>> {
        return [
            {
                customView: "previous",
                position: "left",
                buttonToolTip: "previous"
            },
            {
                customView: "today",
                position: "left",
                customCaption: "today"
            },
            {
                customView: "next",
                position: "left",
                buttonToolTip: "next"
            },
            {
                customView: "title",
                position: "center"
            },
            {
                customView: "day",
                position: "right",
                customCaption: "day"
            },
            {
                customView: "week",
                position: "right",
                customCaption: "week"
            },
            {
                customView: "month",
                position: "right",
                customCaption: "month"
            }
        ];
    }

    private renderAlert(): ReactNode {
        return createElement(Alert, { className: "widget-calendar-alert" }, this.props.alertMessage);
    }

    private onNavigate(date: Date): void {
        this.setState({ date });
    }

    private renderCalendar(): ReactNode {
        const wrapToolbar = (injectedProps: HOCToolbarProps): Function => (toolbarProps: Container.ToolbarProps) =>
            createElement(CustomToolbar as any, { ...injectedProps, ...toolbarProps });

        const props = {
            localizer,
            events: this.props.events,
            allDayAccessor: this.allDayAccessor,
            components: {
                toolbar: wrapToolbar({ customViews: this.getToolbarProps(), onClickToolbarButton: this.onRangeChange })
            },
            onNavigate: this.onNavigateHandler,
            eventPropGetter: this.eventColor,
            date: this.state.date ?? this.props.startPosition,
            defaultView: this.defaultView(),
            formats: this.props.viewOption === "custom" ? this.props.formats : "",
            messages: this.props.viewOption === "custom" ? this.props.messages : "",
            popup: this.props.popup,
            selectable: this.props.enableCreate,
            step: 60,
            showMultiDayTimes: true,
            onRangeChange: this.onRangeChange,
            onSelectEvent: this.onSelectEvent,
            onSelectSlot: this.onSelectSlot,
            views: ["month", "day", "week", "work_week", "month", "agenda"]
        };

        if (this.props.loading) {
            return createElement(CalendarLoader);
        } else if (this.props.enableCreate && this.props.editable === "default") {
            return createElement(DragAndDropCalendar, {
                ...props,
                onEventDrop: this.onEventDrop,
                onEventResize: this.onEventResize
            });
        } else {
            return createElement(BigCalendar, props);
        }
    }

    private defaultView(): Style.View {
        if (
            this.props.viewOption === "standard" &&
            (this.props.defaultView === "work_week" || this.props.defaultView === "agenda")
        ) {
            return "month";
        }

        return this.props.defaultView;
    }

    private getToolbarProps(): Array<Partial<Container.ButtonConfig>> {
        const toolbarProps = this.props.viewOption === "standard" ? this.getDefaultToolbar() : this.props.customViews;

        return toolbarProps.map(customView => ({ ...customView, onClickToolbarButton: this.onRangeChange }));
    }

    private allDayAccessor = (event: Container.ViewOptions) => event.allDay;

    private eventColor = (events: CalendarEvent) => ({ style: { backgroundColor: events.color } });

    private onRangeChange = (date: object) => {
        if (this.props.onRangeChangeAction && date) {
            this.props.onRangeChangeAction(date);
        }
    };

    private onEventDrop = (eventInfo: Container.EventInfo) => {
        if (
            eventInfo.start.getDate() !== eventInfo.event.start.getDate() &&
            this.props.editable === "default" &&
            this.props.onEventDropAction
        ) {
            this.props.onEventDropAction(eventInfo);
        }
    };

    private onEventResize = (_resizeType: string, eventInfo: Container.EventInfo) => {
        if (
            (eventInfo.start.getDate() !== eventInfo.event.start.getDate() ||
                eventInfo.end.getDate() !== eventInfo.event.end.getDate()) &&
            eventInfo.start.getDate() < eventInfo.end.getDate() &&
            this.props.editable &&
            this.props.onEventResizeAction
        ) {
            this.props.onEventResizeAction(eventInfo);
        }
    };

    private onSelectEvent = (eventInfo: object) => {
        if (this.props.onSelectEventAction) {
            this.props.onSelectEventAction(eventInfo);
        }
    };

    private onSelectSlot = (slotInfo: object) => {
        if (this.props.onSelectSlotAction) {
            this.props.onSelectSlotAction(slotInfo);
        }
    };
}

export { Calendar, Calendar as MyCalendar };
