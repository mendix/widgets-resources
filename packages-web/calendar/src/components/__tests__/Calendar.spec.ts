import { shallow, ShallowWrapper } from "enzyme";
import { createElement } from "react";

import { CalendarProps, MyCalendar } from "../Calendar";
import { SizeContainer, SizeProps } from "../SizeContainer";

interface Window {
    mx: any;
}

declare const window: Window;

describe("Calendar", () => {
    const renderCalendar = (props: CalendarProps): ShallowWrapper<CalendarProps, any> =>
        shallow(createElement(MyCalendar, props));
    const renderCalendarSizeComponent = (props: SizeProps): ShallowWrapper<SizeProps, any> =>
        shallow(createElement(SizeContainer, props));
    const sizeProps: SizeProps = {
        className: "widget-calendar",
        widthUnit: "pixels",
        width: 100,
        heightUnit: "pixels",
        height: 100
    };
    const calendarProps: CalendarProps = {
        customViews: [],
        defaultView: "month",
        enableCreate: true,
        events: [],
        formats: {},
        height: 580,
        heightUnit: "pixels",
        messages: [],
        popup: true,
        editable: "default",
        style: {},
        viewOption: "standard",
        width: 100,
        widthUnit: "percentage"
    };

    beforeAll(() => {
        (window as any).mx = {
            session: {
                sessionData: {
                    locale: { firstDayOfWeek: 1 }
                }
            },
            parser: {
                formatValue: () => jest.fn()
            }
        };
    });

    it("renders the structure correctly", () => {
        const calendar = renderCalendarSizeComponent(sizeProps);

        expect(calendar).toMatchSnapshot();
    });

    it("should render a structure correctly with pixels", () => {
        const calendar = renderCalendarSizeComponent(sizeProps);

        expect(calendar).toMatchSnapshot();
    });

    it("should render a structure correctly with percentage", () => {
        const calendar = renderCalendarSizeComponent(sizeProps);
        calendar.setProps({
            heightUnit: "percentageOfWidth",
            widthUnit: "percentage"
        });

        expect(calendar).toMatchSnapshot();
    });

    it("should render a structure correctly with percentage of parent", () => {
        const calendar = renderCalendarSizeComponent(sizeProps);
        calendar.setProps({
            heightUnit: "percentageOfParent",
            widthUnit: "percentage"
        });

        expect(calendar).toMatchSnapshot();
    });

    describe("event handler", () => {
        it("#onSelectSlot() calls the parent onSelectSlot handler", () => {
            calendarProps.onSelectSlotAction = jest.fn();
            const calendar = renderCalendar(calendarProps);
            (calendar.instance() as any).onSelectSlot();

            expect(calendarProps.onSelectSlotAction).toHaveBeenCalled();
        });

        it("#onSelectEvent() calls the parent onSelectEvent handler", () => {
            calendarProps.onSelectEventAction = jest.fn();
            const calendar = renderCalendar(calendarProps);
            (calendar.instance() as any).onSelectEvent();

            expect(calendarProps.onSelectEventAction).toHaveBeenCalled();
        });

        it("#onEventDrop() calls the parent onEventDrop handler when the dates are different", () => {
            const eventInfo = {
                start: new Date(),
                event: {
                    start: new Date(new Date().valueOf() + 1000 * 3600 * 24)
                }
            };
            calendarProps.onEventDropAction = jest.fn();
            const calendar = renderCalendar(calendarProps);
            (calendar.instance() as any).onEventDrop(eventInfo);

            expect(calendarProps.onEventDropAction).toHaveBeenCalled();
        });

        // it("#onEventResize() calls the parent onEventResize handler when the start date, end date or both dates have changed", () => {
        //     const resizeType = "drop";
        //     const startDate = new Date();
        //     startDate.setDate(startDate.getDate() - 2);
        //     const initialEndDate = new Date();
        //     initialEndDate.setDate(initialEndDate.getDate() - 1);
        //     const eventInfo = {
        //         start: startDate,
        //         end: new Date(),
        //         event: {
        //             start: startDate,
        //             end: initialEndDate
        //         }
        //     };
        //     calendarProps.onEventResizeAction = jest.fn();
        //     const calendar = renderCalendar(calendarProps);
        //     (calendar.instance() as any).onEventResize(resizeType, eventInfo);
        //
        //     expect(calendarProps.onEventResizeAction).toHaveBeenCalled();
        // });
    });

    afterAll(() => {
        (window.mx as any) = undefined;
    });
});
