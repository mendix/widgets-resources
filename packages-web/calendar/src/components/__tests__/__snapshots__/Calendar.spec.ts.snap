// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar renders the structure correctly 1`] = `
<div
  className="widget-calendar size-box"
  style={
    Object {
      "height": "100px",
      "position": "relative",
      "width": "100px",
    }
  }
>
  <div
    className="size-box-inner"
    style={
      Object {
        "bottom": "0",
        "left": "0",
        "position": "absolute",
        "right": "0",
        "top": "0",
      }
    }
  />
</div>
`;

exports[`Calendar should render a structure correctly with percentage 1`] = `
<div
  className="widget-calendar size-box"
  style={
    Object {
      "height": "auto",
      "paddingBottom": "100%",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <div
    className="size-box-inner"
    style={
      Object {
        "bottom": "0",
        "left": "0",
        "position": "absolute",
        "right": "0",
        "top": "0",
      }
    }
  />
</div>
`;

exports[`Calendar should render a structure correctly with percentage of parent 1`] = `
<div
  className="widget-calendar size-box"
  style={
    Object {
      "height": "100%",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <div
    className="size-box-inner"
    style={
      Object {
        "bottom": "0",
        "left": "0",
        "position": "absolute",
        "right": "0",
        "top": "0",
      }
    }
  />
</div>
`;

exports[`Calendar should render a structure correctly with pixels 1`] = `
<div
  className="widget-calendar size-box"
  style={
    Object {
      "height": "100px",
      "position": "relative",
      "width": "100px",
    }
  }
>
  <div
    className="size-box-inner"
    style={
      Object {
        "bottom": "0",
        "left": "0",
        "position": "absolute",
        "right": "0",
        "top": "0",
      }
    }
  />
</div>
`;
